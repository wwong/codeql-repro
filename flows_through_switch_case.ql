/**
 * @id ww/switch-case-repro
 * @kind path-problem
 * @problem.severity info
 */

import java
import semmle.code.java.dataflow.DataFlow

module StringFlowConfig implements DataFlow::ConfigSig {
    predicate isSource(DataFlow::Node source) {
        source.asExpr() instanceof StringLiteral and
        source.getEnclosingCallable().hasName("main")
    }

    predicate isSink(DataFlow::Node sink) {
        sink.getEnclosingCallable().hasName("sayHi")
    }

    // predicate isAdditionalFlowStep(DataFlow::Node n1, DataFlow::Node n2) {
    //     n1.asParameter().getAnAccess() = n2.asExpr()
    // }
}

module StringFlow = DataFlow::Global<StringFlowConfig>;
import StringFlow::PathGraph

from StringFlow::PathNode source, StringFlow::PathNode sink
where StringFlow::flowPath(source, sink)
select sink.getNode(), source, sink, sink.getNode().getEnclosingCallable().getName()
