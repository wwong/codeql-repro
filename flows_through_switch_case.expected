edges
| HelloWorld.java:5:32:5:39 | "Semmle" : String | HelloWorld.java:8:24:8:34 | name : String |
| HelloWorld.java:8:24:8:34 | name : String | HelloWorld.java:21:15:21:18 | name : String |
| HelloWorld.java:21:15:21:18 | name : String | HelloWorld.java:24:16:24:26 | name |
| HelloWorld.java:21:15:21:18 | name : String | HelloWorld.java:24:16:24:26 | name : String |
| HelloWorld.java:24:16:24:26 | name : String | HelloWorld.java:25:37:25:40 | name |
nodes
| HelloWorld.java:5:32:5:39 | "Semmle" : String | semmle.label | "Semmle" : String |
| HelloWorld.java:8:24:8:34 | name : String | semmle.label | name : String |
| HelloWorld.java:21:15:21:18 | name : String | semmle.label | name : String |
| HelloWorld.java:24:16:24:26 | name | semmle.label | name |
| HelloWorld.java:24:16:24:26 | name : String | semmle.label | name : String |
| HelloWorld.java:25:37:25:40 | name | semmle.label | name |
subpaths
#select
| HelloWorld.java:24:16:24:26 | name | HelloWorld.java:5:32:5:39 | "Semmle" : String | HelloWorld.java:24:16:24:26 | name | sayHi |
| HelloWorld.java:25:37:25:40 | name | HelloWorld.java:5:32:5:39 | "Semmle" : String | HelloWorld.java:25:37:25:40 | name | sayHi |
